packages/flutter_tools/lib/src/preview_device.dart

```diff
@@ -28,30 +29,107 @@ BundleBuilder _defaultBundleBuilder() {
   return BundleBuilder();
 }
 
+class PreviewDeviceDiscovery extends DeviceDiscovery {
+  PreviewDeviceDiscovery({
+    required Platform platform,
+    required Artifacts artifacts,
+    required FileSystem fileSystem,
+    required Logger logger,
+    required ProcessManager processManager,
+    required FeatureFlags featureFlags,
+  }) : _artifacts = artifacts,
+       _logger = logger,
+       _processManager = processManager,
+       _fileSystem = fileSystem,
+       _platform = platform,
+       _features = featureFlags;
+
+  final Platform _platform;
+  final Artifacts _artifacts;
+  final Logger _logger;
+  final ProcessManager _processManager;
+  final FileSystem _fileSystem;
+  final FeatureFlags _features;
+
+  @override
+  bool get canListAnything => _platform.isWindows;
+
+  @override
+  Future<List<Device>> devices({
+    Duration? timeout,
+    DeviceDiscoveryFilter? filter,
+  }) async {
+    if (!_platform.isWindows) {
+      // we need to early return before looking up the path of [Artifact.flutterPreviewDevice].
+      return const <Device>[];
+    }
+    final File previewBinary = _fileSystem.file(_artifacts.getArtifactPath(Artifact.flutterPreviewDevice));
+    if (!previewBinary.existsSync()) {
+      return const <Device>[];
+    }
+    final PreviewDevice device = PreviewDevice(
+      artifacts: _artifacts,
+      fileSystem: _fileSystem,
+      logger: _logger,
+      processManager: _processManager,
+      previewBinary: previewBinary,
+    );
+    final bool matchesRequirements;
+    if (!_features.isPreviewDeviceEnabled) {
+      matchesRequirements = false;
+    } else if (filter == null) {
+      matchesRequirements = true;
+    } else {
+      matchesRequirements = await filter.matchesRequirements(device);
+    }
+    return <Device>[
+      if (matchesRequirements)
+        device,
+    ];
+  }
+
+  @override
+  Future<List<Device>> discoverDevices({
+    Duration? timeout,
+    DeviceDiscoveryFilter? filter,
+  }) {
+    return devices();
+  }
+
+  @override
+  bool get supportsPlatform => true;
+
+  @override
+  List<String> get wellKnownIds => <String>['preview'];
+}
+
 /// A device type that runs a prebuilt desktop binary alongside a locally compiled kernel file.
 ///
 /// This could be used to support debug local development without plugins on machines that
 /// have not completed the SDK setup. These features are not fully implemented and the
 /// device is not currently discoverable.
```

Comment:
Should we update this documentation to say "device is currently discoverable with feature flags turned on?"

Reply:
:+1:

Reply:
done